### PR TITLE
feature based UAV image alignment experiment 

### PR DESCRIPTION
**Pull request by AnNyiiik**

### CODE REVIEW

File: Dockerfile
Comment (vnmsklnk) [inline Dockerfile:1]: Why do we need Docker? In this case, one would like to see instructions for creating a proper Python environment. There should be a `requirements.txt` file, as well as specifications of which versions of Python are suitable

File: README.md [hunk 1]
```diff
@@ -1 +1,111 @@
-# AlignmentOfUAVImages
\ No newline at end of file
+# AlignmentOfUAVImages
+### Description
+This tool is designed to calculate UAV pixel coordinates in the world coordinate system using the corresponded satellite
```
Comment (vnmsklnk) [inline README.md]: This is coordinate system of the satellite image, not a world coordinate system, isn't it?

File: README.md [hunk 2]
```diff
@@ -1 +1,111 @@
...
+image. The transformation between pair of pictures calculated using key points pairs extracted from both of images. 
+![visualization](https://github.com/AnNyiiik/HWThirdTerm/assets/114094098/b6596732-c49f-47cc-baa8-01f9c97bfd31)
+### Usage
+1. Clone the repository running `git clone git@github.com:AnNyiiik/AlignmentOfUAVImages.git` on a terminal.
+1. Add your data to the repository root folder  [example of data](exampleOfData).
+1. Build docker image running `docker build -t alignment .`
+1. Create a container using image id: `docker start container id`.
+1. Run the command to execute the experiment: `python3 experiment.py path\to\UAV\KeyPoints path\to\satellite\KeyPoints
```
Comment (vnmsklnk) [inline README.md]: It's an MacOS path style, so maybe `PATH_TO_UAV_KEYPOINTS` will be look better

File: README.md [hunk 3]
```diff
@@ -1 +1,111 @@
...
+1. Build docker image running `docker build -t alignment .`
+1. Create a container using image id: `docker start container id`.
+1. Run the command to execute the experiment: `python3 experiment.py path\to\UAV\KeyPoints path\to\satellite\KeyPoints 
+   path\to\UAV\image path\to\satellite\image path\to\meta\data`.
+1. The image with matched key points, hist of reprojection error, file with the reprojection error value, 
+   standard deviation of error and pixel coordinates are:
+   * image_matches.png
+   * reprojection_error.png
+   * results 
+### Experimental results
```
Comment (vnmsklnk) [inline README.md:17]: The table looks scary. We should discuss how to change it

File: alignment.py [hunk 4]
```diff
@@ -0,0 +1,67 @@
+import cv2 as cv
+import numpy as np
+import math
+
+def readKeyPointsFromFile(path):
```
Comment (vnmsklnk) [inline alignment.py]: It's a Python, so snake_case should be everywhere instead of camelCase

File: alignment.py [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    with open(path, 'r') as f:
+        lines = [line.rstrip() for line in f]
+    f.close()
```
Comment (vnmsklnk) [inline alignment.py]: `f.close()` is unnecessary because you use `with` statement

File: alignment.py [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+    height, width = imageSource.shape[:2]
+    pixel = cv.KeyPoint(pixel[0], pixel[1], 1)
+    if (pixel.pt[0] < 0 or pixel.pt[0] > width):
+        return
+    if (pixel.pt[1] < 0 or pixel.pt[1] > height):
+        return
+    height, width = destinationImage.shape[:2]
+    pointInDestinationImage = np.array(homographyMatrix).dot(np.array([pixel.pt[0], pixel.pt[1], 1]))
+    print(pointInDestinationImage[0], pointInDestinationImage[1])
+    lat = coordinatesOfTopLeft[0] + float(pointInDestinationImage[1]) / height * float((coordinatesOfBottomRight[0] -
```
Comment (vnmsklnk) [inline alignment.py]: Very long lines, please use `black` formatter

File: alignment.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+        source[i, 1] = keyPointsAerial[matches[i].queryIdx].pt[1]
+        reference[i, 0] = keyPointsSatellite[matches[i].trainIdx].pt[0]
+        reference[i, 1] = keyPointsSatellite[matches[i].trainIdx].pt[1]
+    H, _ = cv.findHomography(source, reference, cv.RANSAC)
+    return H
+
+def findMatches(pairIndexes):
+    matches = []
+    for pair in pairIndexes:
+        matches.append(cv.DMatch(pair, pair, 1.7976931348623157e+308))
```
Comment (vnmsklnk) [inline alignment.py]: magic constant

File: alignment.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+
+def findHomographyTransform(keyPointsAerial, keyPointsSatellite, matches):
+    source = np.empty((len(matches), 2), dtype=np.float32)
+    reference = np.empty((len(matches), 2), dtype=np.float32)
+    for i in range(len(matches)):
+        source[i, 0] = keyPointsAerial[matches[i].queryIdx].pt[0]
+        source[i, 1] = keyPointsAerial[matches[i].queryIdx].pt[1]
+        reference[i, 0] = keyPointsSatellite[matches[i].trainIdx].pt[0]
+        reference[i, 1] = keyPointsSatellite[matches[i].trainIdx].pt[1]
+    H, _ = cv.findHomography(source, reference, cv.RANSAC)
```
Comment (vnmsklnk) [inline alignment.py]: very difficult, we can just pass lists of keypoints pairs to the `findHomography`

File: alignment.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        matches.append(cv.DMatch(pair, pair, 1.7976931348623157e+308))
+    return matches
+
+def findReprojectionError(points_truth, points_under_homography):
+    if (len(points_truth) == 0):
+        return
+    error = 0
+    declines = []
+    for i in range(len(points_truth)):
+        decline = math.sqrt(math.pow(points_truth[i][0] - points_under_homography[i][0], 2) + \
```
Comment (vnmsklnk) [inline alignment.py]: Please, check how to use `np.linalg.norm` for calculating distance

File: experiment.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
+import matplotlib.pyplot as plt
+import sys
+from alignment import *
+import statistics
```
Comment (vnmsklnk) [inline experiment.py]: 1. Imports should be sorted
2. Please, avoid star imports

File: experiment.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+#parse arguments
```
Comment (vnmsklnk) [inline experiment.py]: let's use `argparse` for parsing

File: experiment.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    # count pixel coordinates
+    lat, lon = countPixelCoordinates(aerialImage, (X, Y), satelliteImage, H, (lat_left, lon_left),
+                                     (lat_right, lon_right))
+
+    # save hist of reprojection error and its value to the experiment results folder
+    f = open("results", 'w')
+    f.write("reprojection error: " + str(error))
+    f.write("standard deviation of reprojection error: " + str(statistics.stdev(declines)))
+    f.write("pixel coordinates (lat, lon): " + str(lat) + " " + str(lon))
+    f.close()
```
Comment (vnmsklnk) [inline experiment.py]: Let's use `with` here, then `open` and `close` is unnecessary

File: requirements.txt
Comment (vnmsklnk) [inline requirements.txt:1]: Requirements file should contain concrete versions of necessary packages

File: experiment.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    cv.imwrite("image_matches.png", imageMatches)
+
+    # compute homography transform
+    H = findHomographyTransform(keyPointsAerial, keyPointsSatellite, matches)
+
+    # count reprojection error
+    points_under_homography = [[0] * 2 for j in range(len(keyPointsAerial))]
+    truth_points = [[0] * 2 for j in range(len(keyPointsAerial))]
+    for j in range(len(keyPointsAerial)):
+        destPoint = np.array(H).dot(np.array([keyPointsAerial[matches[j].queryIdx].pt[0],
```
Comment (vnmsklnk) [inline experiment.py]: it's a copypaste from `countPixelCoordinates`

File: alignment.py [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+    with open(path, 'r') as f:
+        lines = [line.rstrip() for line in f]
+    f.close()
+    keyPoints = list()
+    for line in lines:
+        point = list(map(int, line.split()))
+        keyPoints.append(cv.KeyPoint(point[0], point[1], 1))
+    return keyPoints
+
+def countPixelCoordinates(imageSource, pixel, destinationImage, homographyMatrix, coordinatesOfTopLeft,
```
Comment (vnmsklnk) [inline alignment.py]: it's to better to call it `calculate`

**Review by vnmsklnk — CHANGES_REQUESTED** [review]

File: Dockerfile
Comment (AnNyiiik) [reply Dockerfile:1]: reprojection_error

Comment (AnNyiiik) [reply Dockerfile:1]: reprojection_error_esri
google_sat_error

File: alignment.py (hunk 14 above)
Comment (AnNyiiik) [reply alignment.py]: reprojection_error_esri
google_satellite_error_pandas

File: alignment.py [hunk 15]
```diff
@@ -0,0 +1,109 @@
+import cv2 as cv
+import numpy as np
+import math
+
+distance_for_match = 1.7976931348623157e308
```
Comment (vnmsklnk) [inline alignment.py]: I'm not sure where this constant came from

File: alignment.py [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+    return key_points
+
+
+def calculate_point_in_destination_image(image_source, pixel, homography_matrix):
+    height, width = image_source.shape[:2]
+    if pixel.pt[0] < 0 or pixel.pt[0] > width:
+        return
+    if pixel.pt[1] < 0 or pixel.pt[1] > height:
+        return
+    point_in_destination_image = np.array(homography_matrix).dot(
```
Comment (vnmsklnk) [inline alignment.py]: Let's use `perspectiveTransform` from OpenCV ([example](https://stackoverflow.com/a/55055829)).
It seems that there is an error in your calculations now, you can check [here](https://stackoverflow.com/a/9275631).

File: alignment.py [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    )
+    cv.drawMatches(
+        aerial_image,
+        key_points_aerial,
+        satellite_image,
+        key_points_satellite,
+        matches,
+        img_matches,
+        flags=cv.DrawMatchesFlags_NOT_DRAW_SINGLE_POINTS,
+    )
```
Comment (vnmsklnk) [inline alignment.py]:
```suggestion
    img_matches = cv.drawMatches(
        aerial_image,
        key_points_aerial,
        satellite_image,
        key_points_satellite,
        matches,
        None,
    )
```

File: alignment.py [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+
+def find_homography_transform(key_points_aerial, key_points_satellite):
+    source = [
+        [key_points_aerial[i].pt[0], key_points_aerial[i].pt[1]]
+        for i in range(len(key_points_aerial))
+    ]
+    dest = [
+        [key_points_satellite[i].pt[0], key_points_satellite[i].pt[1]]
+        for i in range(len(key_points_aerial))
+    ]
```
Comment (vnmsklnk) [inline alignment.py]:
```suggestion
    source = [[key_point.pt[0], key_point.pt[1]] for key_point in key_points_aerial]
    dest = [[key_point.pt[0], key_point.pt[1]] for key_point in key_points_satellite]
```

File: experiment.py [hunk 19]
```diff
@@ -0,0 +1,101 @@
+import alignment
+import argparse
+import cv2 as cv
+import matplotlib.pyplot as plt
+import numpy as np
+import statistics
+import sys
+
+parser = argparse.ArgumentParser()
+parser.add_argument("path_to_UAV_key_points", type=str, nargs=1)
```
Comment (vnmsklnk) [inline experiment.py]: `nargs` is unnecessary

File: experiment.py [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+
+# read pixel coordinates of UAV image to count its (lat, lon) coordinates, read satellite image corners' coordinates
+with open(args.path_to_meta_data[0], "r") as f:
+    lat_left, lon_left = map(float, f.readline().split())
+    lat_right, lon_right = map(float, f.readline().split())
+    X, Y = map(int, f.readline().split())
+
+# find and draw matches
+pair_indexes = [j for j in range(len(key_points_aerial))]
+matches = alignment.find_matches(pair_indexes)
```
Comment (vnmsklnk) [inline experiment.py]:
```suggestion
matches = [cv.DMatch(i, i, 1) for i in range(len(key_points_aerial))]
```
And you don't need `find_matches` then

File: experiment.py [hunk 21]
```diff
@@ -0,0 +1,101 @@
...
+image_matches = alignment.draw_matches(
+    matches, aerial_image, satellite_image, key_points_aerial, key_points_satellite
+)
+cv.imwrite("image_matches.png", image_matches)
+
+# compute homography transform
+H = alignment.find_homography_transform(key_points_aerial, key_points_satellite)
+
+# count reprojection error
+points_under_homography = [[0] * 2 for j in range(len(key_points_aerial))]
```
Comment (vnmsklnk) [inline experiment.py]: It's better to use `np.zeros` for this. And even better to adapt `calculate_point_in_destination_image` for working with multiple points

File: requirements.txt [hunk 22]
```diff
@@ -0,0 +1,4 @@
+matplotlib>=3.7.3
```
Comment (vnmsklnk) [inline requirements.txt]: There should be strict equality here, in case some package gets updated and something breaks

**Review by vnmsklnk — CHANGES_REQUESTED** [review]

File: requirements.txt (hunk 22 above)
Comment (AnNyiiik) [reply requirements.txt]: reprojection_error_for_each_pair

reprojection_error_for_each_pair

Comment (AnNyiiik) [reply requirements.txt]: Figure_1

**Pull request merged**